lti_consumer/tests/unit/test_lti_xblock.py

```diff
@@ -1532,3 +1534,152 @@ def test_access_token(self):
+            urllib.parse.urlencode({
+                "grant_type": "client_credentials",
+                "client_assertion_type": "something",
+                "client_assertion": jwt,
+                "scope": "",
+            }),
+            'POST'
+        )
+        request.content_type = 'application/x-www-form-urlencoded'
+
+        jwks = KEYS()
+        jwks._keys = [self.key]  # pylint: disable=protected-access
+        load_jwks_from_url.return_value = jwks
+
+        response = self.xblock.lti_1p3_access_token(request)
+        load_jwks_from_url.assert_called_once_with('http://tool.example/keyset')
+        self.assertEqual(response.status_code, 200)
+
+    @patch('lti_consumer.lti_1p3.key_handlers.load_jwks_from_url')
+    def test_access_token_using_keyset_url_with_empty_keys(self, load_jwks_from_url):
+        """
+        Test request with valid JWT, where the provider's keyset URL returns an empty list of keys.
+        """
+        self.xblock.lti_1p3_tool_public_key = ''
+        self.xblock.lti_1p3_tool_keyset_url = 'http://tool.example/keyset'
+        self.xblock.save()
+
+        jwt = create_jwt(self.key, {})
+        request = make_request(
+            urllib.parse.urlencode({
+                "grant_type": "client_credentials",
+                "client_assertion_type": "something",
+                "client_assertion": jwt,
+                "scope": "",
+            }),
+            'POST'
+        )
+        request.content_type = 'application/x-www-form-urlencoded'
+
+        jwks = KEYS()
+        jwks._keys = []  # pylint: disable=protected-access
+        load_jwks_from_url.return_value = jwks
+
+        response = self.xblock.lti_1p3_access_token(request)
+        load_jwks_from_url.assert_called_once_with('http://tool.example/keyset')
+        self.assertEqual(response.status_code, 400)
+        self.assertEqual(response.json_body, {'error': 'invalid_client'})
+
+    @patch('lti_consumer.lti_1p3.key_handlers.load_jwks_from_url')
+    def test_access_token_using_keyset_url_with_wrong_keys(self, load_jwks_from_url):
+        """
+        Test request with valid JWT, where the provider's keyset URL returns wrong keys.
+        """
+        self.xblock.lti_1p3_tool_public_key = ''
+        self.xblock.lti_1p3_tool_keyset_url = 'http://tool.example/keyset'
+        self.xblock.save()
+
+        jwt = create_jwt(self.key, {})
+        request = make_request(
+            urllib.parse.urlencode({
+                "grant_type": "client_credentials",
+                "client_assertion_type": "something",
+                "client_assertion": jwt,
+                "scope": "",
+            }),
+            'POST'
+        )
+        request.content_type = 'application/x-www-form-urlencoded'
+
+        rsa_key = RSA.generate(2048)
+        key = RSAKey(key=rsa_key, kid='2')
+
+        jwks = KEYS()
+        jwks._keys = [key]  # pylint: disable=protected-access
+        load_jwks_from_url.return_value = jwks
```

Comment:
@felipetrz You are reusing a lot of variations for this section, can you push them to a separate method? Or at least the parts of it that are re-usable. You can set up patchers directly on the `setUp` method if that makes testing easier.

Reply:
@giovannicimolin I refactored and cleaned the tests up.